File: format_mab.py
"""
Multi-armed bandit (MAB) for format selection: reward estimates, epsilon-greedy selection,
outcome recording, and persistence. AdaEdge-inspired; see IMPLEMENTATION_PLAN_format_selection.md.
"""

from __future__ import annotations

import json
import os
import random
from pathlib import Path
from typing import Any, Dict, Literal, Optional

from format_selector import (
    FormatName,
    HintsDict,
    OptimizationTarget,
    SelectionContext,
    select_format,
    select_format_with_hints,
)

# Outcome record: context + choice + observed metrics (for history and reward)
OutcomeDict = Dict[str, Any]  # n_rows, n_cols, target, format, bytes, latency_s, time_to_first_rows_s?

# MAB state: per (target, format) -> {sum_reward, count}
MABStateDict = Dict[str, Dict[str, Dict[str, float]]]  # target -> format -> {sum_reward, count}

DEFAULT_HISTORY_PATH = Path("results/format_selection_history.jsonl")
DEFAULT_MAB_STATE_PATH = Path("results/format_mab_state.json")
FORMATS: tuple[FormatName, ...] = (
    "json",
    "parquet_blob",
    "parquet_stream",
    "arrow_ipc_blob",
    "arrow_ipc_stream",
    "text_inline",
    "raw_blob",
    "gzip_blob",
)


def reward_from_outcome(outcome: OutcomeDict, target: OptimizationTarget) -> float:
    """
    Compute reward from outcome (to maximize). Higher is better.
    min_bytes -> -bytes; min_latency -> -latency_s; min_time_to_first_rows -> -time_to_first_rows_s.
    """
    if target == OptimizationTarget.MIN_BYTES:
        b = outcome.get("bytes")
        if b is None:
            return 0.0
        return -float(b)
    if target == OptimizationTarget.MIN_LATENCY:
        s = outcome.get("latency_s")
        if s is None:
            return 0.0
        return -float(s)
    if target == OptimizationTarget.MIN_TIME_TO_FIRST_ROWS:
        ttfr = outcome.get("time_to_first_rows_s")
        if ttfr is not None:
            return -float(ttfr)
        # Fallback: use latency as proxy
        s = outcome.get("latency_s")
        if s is not None:
            return -float(s)
        return 0.0
    return 0.0


def _state_key(target: OptimizationTarget, format_name: FormatName) -> tuple[str, str]:
    return (target.value, format_name)


def load_mab_state(path: Optional[os.PathLike[str]] = None) -> MABStateDict:
    """Load MAB state from JSON; return empty dict if file missing or invalid."""
    p = Path(path) if path is not None else DEFAULT_MAB_STATE_PATH
    if not p.is_file():
        return {}
    try:
        with open(p) as f:
            data = json.load(f)
    except (json.JSONDecodeError, OSError):
        return {}
    if not isinstance(data, dict):
        return {}
    # Normalize: data is {target: {format: {sum_reward, count}}}
    out: MABStateDict = {}
    for target_val, formats_dict in data.items():
        if not isinstance(formats_dict, dict):
            continue
        out[target_val] = {}
        for fmt, stats in formats_dict.items():
            if isinstance(stats, dict) and "sum_reward" in stats and "count" in stats:
                out[target_val][fmt] = {
                    "sum_reward": float(stats["sum_reward"]),
                    "count": float(stats["count"]),
                }
    return out


def save_mab_state(state: MABStateDict, path: Optional[os.PathLike[str]] = None) -> None:
    """Persist MAB state to JSON."""
    p = Path(path) if path is not None else DEFAULT_MAB_STATE_PATH
    p.parent.mkdir(parents=True, exist_ok=True)
    with open(p, "w") as f:
        json.dump(state, f, indent=2)


def select_format_with_mab(
    context: SelectionContext,
    hints: Optional[HintsDict],
    mab_state: Optional[MABStateDict],
    *,
    epsilon: Optional[float] = None,
) -> FormatName:
    """
    Epsilon-greedy format selection using MAB state. If mab_state is None or has no
    data for this target, fall back to select_format_with_hints (or select_format).
    """
    if mab_state is None:
        if hints:
            return select_format_with_hints(context, hints)
        return select_format(context)

    target_val = context.target.value
    state_for_target = mab_state.get(target_val)
    if not state_for_target:
        if hints:
            return select_format_with_hints(context, hints)
        return select_format(context)

    # Check if we have any observations
    has_any = any(
        state_for_target.get(fmt, {}).get("count", 0) > 0 for fmt in FORMATS
    )
    if not has_any:
        if hints:
            return select_format_with_hints(context, hints)
        return select_format(context)

    eps = epsilon if epsilon is not None else float(os.environ.get("FORMAT_SELECT_EPSILON", "0.1"))
    if random.random() < eps:
        return random.choice(list(FORMATS))

    # Greedy: argmax Q(format) = sum_reward / count
    best_format: Optional[FormatName] = None
    best_q = float("-inf")
    for fmt in FORMATS:
        stats = state_for_target.get(fmt, {})
        count = stats.get("count", 0) or 0
        if count <= 0:
            continue
        q = (stats.get("sum_reward", 0) or 0) / count
        if q > best_q:
            best_q = q
            best_format = fmt

    if best_format is not None:
        return best_format
    if hints:
        return select_format_with_hints(context, hints)
    return select_format(context)


def record_outcome(
    context: SelectionContext,
    format_used: FormatName,
    outcome: OutcomeDict,
    mab_state: Optional[MABStateDict] = None,
    history_path: Optional[os.PathLike[str]] = None,
) -> None:
    """
    Append outcome to history file and update MAB state (sum_reward, count) for (target, format_used).
    """
    # Build record for history
    record: OutcomeDict = {
        "n_rows": context.n_rows,
        "n_cols": context.n_cols,
        "target": context.target.value,
        "format": format_used,
        "bytes": outcome.get("bytes"),
        "latency_s": outcome.get("latency_s"),
        "time_to_first_rows_s": outcome.get("time_to_first_rows_s"),
    }

    # Append to history file
    p = Path(history_path) if history_path is not None else DEFAULT_HISTORY_PATH
    p.parent.mkdir(parents=True, exist_ok=True)
    with open(p, "a") as f:
        f.write(json.dumps(record) + "\n")

    # Update MAB state
    if mab_state is not None:
        reward = reward_from_outcome(outcome, context.target)
        target_val = context.target.value
        if target_val not in mab_state:
            mab_state[target_val] = {}
        if format_used not in mab_state[target_val]:
            mab_state[target_val][format_used] = {"sum_reward": 0.0, "count": 0.0}
        mab_state[target_val][format_used]["sum_reward"] += reward
        mab_state[target_val][format_used]["count"] += 1.0


def mab_enabled() -> bool:
    """True if FORMAT_SELECT_MAB env is set to enable MAB selection."""
    return os.environ.get("FORMAT_SELECT_MAB", "").strip() in ("1", "true", "yes")

File: java-server/src/main/java/com/multimodal/mcp/format/FormatMab.java
package com.multimodal.mcp.format;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Random;
import java.util.concurrent.ThreadLocalRandom;

import org.springframework.stereotype.Component;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.multimodal.mcp.util.Env;

/**
 * Multi-armed bandit (MAB) for format selection: reward estimates, epsilon-greedy selection,
 * outcome recording, and persistence.
 */
@Component
public class FormatMab {

    public static final Path DEFAULT_HISTORY_PATH = Path.of("results/format_selection_history.jsonl");
    public static final Path DEFAULT_MAB_STATE_PATH = Path.of("results/format_mab_state.json");

    public static final List<FormatName> FORMATS = List.of(
            FormatName.JSON,
            FormatName.PARQUET_BLOB,
            FormatName.PARQUET_STREAM,
            FormatName.ARROW_IPC_BLOB,
            FormatName.ARROW_IPC_STREAM,
            FormatName.TEXT_INLINE,
            FormatName.RAW_BLOB,
            FormatName.GZIP_BLOB
    );

    private final FormatSelector formatSelector;
    private final ObjectMapper objectMapper = new ObjectMapper();
    private final Random random = ThreadLocalRandom.current();

    public FormatMab(FormatSelector formatSelector) {
        this.formatSelector = formatSelector;
    }

    public double rewardFromOutcome(Map<String, Object> outcome, OptimizationTarget target) {
        if (target == OptimizationTarget.MIN_BYTES) {
            Object bytes = outcome.get("bytes");
            if (bytes == null) {
                return 0.0;
            }
            return -toDouble(bytes);
        }
        if (target == OptimizationTarget.MIN_LATENCY) {
            Object latency = outcome.get("latency_s");
            if (latency == null) {
                return 0.0;
            }
            return -toDouble(latency);
        }
        if (target == OptimizationTarget.MIN_TIME_TO_FIRST_ROWS) {
            Object ttfr = outcome.get("time_to_first_rows_s");
            if (ttfr != null) {
                return -toDouble(ttfr);
            }
            Object latency = outcome.get("latency_s");
            if (latency != null) {
                return -toDouble(latency);
            }
            return 0.0;
        }
        return 0.0;
    }

    public Map<String, Map<String, Map<String, Double>>> loadMabState(Path path) {
        Path resolved = path != null ? path : defaultMabStatePath();
        if (!Files.isRegularFile(resolved)) {
            return new HashMap<>();
        }
        try {
            Map<String, Object> data = objectMapper.readValue(resolved.toFile(), new TypeReference<>() {
            });
            if (data == null) {
                return new HashMap<>();
            }
            Map<String, Map<String, Map<String, Double>>> out = new HashMap<>();
            for (Map.Entry<String, Object> targetEntry : data.entrySet()) {
                if (!(targetEntry.getValue() instanceof Map<?, ?> formatsDict)) {
                    continue;
                }
                Map<String, Map<String, Double>> formats = new HashMap<>();
                for (Map.Entry<?, ?> fmtEntry : formatsDict.entrySet()) {
                    if (fmtEntry.getValue() instanceof Map<?, ?> statsMap) {
                        Object sumReward = statsMap.get("sum_reward");
                        Object count = statsMap.get("count");
                        if (sumReward != null && count != null) {
                            Map<String, Double> stats = new HashMap<>();
                            stats.put("sum_reward", toDouble(sumReward));
                            stats.put("count", toDouble(count));
                            formats.put(String.valueOf(fmtEntry.getKey()), stats);
                        }
                    }
                }
                out.put(targetEntry.getKey(), formats);
            }
            return out;
        } catch (IOException e) {
            return new HashMap<>();
        }
    }

    public void saveMabState(Map<String, Map<String, Map<String, Double>>> state, Path path) throws IOException {
        Path resolved = path != null ? path : defaultMabStatePath();
        Files.createDirectories(resolved.getParent());
        objectMapper.writerWithDefaultPrettyPrinter().writeValue(resolved.toFile(), state);
    }

    public FormatName selectFormatWithMab(
            SelectionContext context,
            Map<String, Object> hints,
            Map<String, Map<String, Map<String, Double>>> mabState,
            Double epsilon) {
        if (mabState == null) {
            return hints != null
                    ? formatSelector.selectFormatWithHints(context, hints)
                    : formatSelector.selectFormat(context);
        }

        String targetVal = context.getTarget().value();
        Map<String, Map<String, Double>> stateForTarget = mabState.get(targetVal);
        if (stateForTarget == null || stateForTarget.isEmpty()) {
            return hints != null
                    ? formatSelector.selectFormatWithHints(context, hints)
                    : formatSelector.selectFormat(context);
        }

        boolean hasAny = FORMATS.stream()
                .anyMatch(fmt -> {
                    Map<String, Double> stats = stateForTarget.get(fmt.value());
                    return stats != null && stats.getOrDefault("count", 0.0) > 0;
                });
        if (!hasAny) {
            return hints != null
                    ? formatSelector.selectFormatWithHints(context, hints)
                    : formatSelector.selectFormat(context);
        }

        double eps = epsilon != null ? epsilon : Env.getDouble("FORMAT_SELECT_EPSILON", 0.1);
        if (random.nextDouble() < eps) {
            return FORMATS.get(random.nextInt(FORMATS.size()));
        }

        FormatName bestFormat = null;
        double bestQ = Double.NEGATIVE_INFINITY;
        for (FormatName fmt : FORMATS) {
            Map<String, Double> stats = stateForTarget.get(fmt.value());
            if (stats == null) {
                continue;
            }
            double count = stats.getOrDefault("count", 0.0);
            if (count <= 0) {
                continue;
            }
            double q = stats.getOrDefault("sum_reward", 0.0) / count;
            if (q > bestQ) {
                bestQ = q;
                bestFormat = fmt;
            }
        }

        if (bestFormat != null) {
            return bestFormat;
        }
        return hints != null
                ? formatSelector.selectFormatWithHints(context, hints)
                : formatSelector.selectFormat(context);
    }

    public void recordOutcome(
            SelectionContext context,
            FormatName formatUsed,
            Map<String, Object> outcome,
            Map<String, Map<String, Map<String, Double>>> mabState,
            Path historyPath) throws IOException {
        Map<String, Object> record = new HashMap<>();
        record.put("n_rows", context.getNRows());
        record.put("n_cols", context.getNCols());
        record.put("target", context.getTarget().value());
        record.put("format", formatUsed.value());
        record.put("bytes", outcome.get("bytes"));
        record.put("latency_s", outcome.get("latency_s"));
        record.put("time_to_first_rows_s", outcome.get("time_to_first_rows_s"));

        Path resolvedHistory = historyPath != null ? historyPath : DEFAULT_HISTORY_PATH;
        Files.createDirectories(resolvedHistory.getParent());
        Files.writeString(resolvedHistory, objectMapper.writeValueAsString(record) + System.lineSeparator(),
                java.nio.file.StandardOpenOption.CREATE, java.nio.file.StandardOpenOption.APPEND);

        if (mabState != null) {
            double reward = rewardFromOutcome(outcome, context.getTarget());
            String targetVal = context.getTarget().value();
            mabState.computeIfAbsent(targetVal, k -> new HashMap<>());
            Map<String, Map<String, Double>> targetState = mabState.get(targetVal);
            targetState.computeIfAbsent(formatUsed.value(), k -> {
                Map<String, Double> stats = new HashMap<>();
                stats.put("sum_reward", 0.0);
                stats.put("count", 0.0);
                return stats;
            });
            Map<String, Double> stats = targetState.get(formatUsed.value());
            stats.put("sum_reward", stats.get("sum_reward") + reward);
            stats.put("count", stats.get("count") + 1.0);
        }
    }

    public boolean mabEnabled() {
        return Env.isTruthy("FORMAT_SELECT_MAB");
    }

    public Path defaultMabStatePath() {
        return Env.getPath("FORMAT_MAB_STATE_PATH", DEFAULT_MAB_STATE_PATH.toString());
    }

    private static double toDouble(Object value) {
        if (value instanceof Number number) {
            return number.doubleValue();
        }
        return Double.parseDouble(String.valueOf(value));
    }
}
